src/items/generics.md

```diff
@@ -43,18 +43,137 @@ struct Ref<'a, T> where T: 'a { r: &'a T }
+3. As a value in any runtime expression in the body of any functions in the
+   item.
+4. As a parameter to any type used in the body of any functions in the item.
+5. As a part of the type of any fields in the item.
+
+```rust
+// Examples where const generic parameters can be used.
+
+// Used in the signature of the item itself.
+fn foo<const N: usize>(arr: [i32; N]) {
+    // Used as a type within a function body.
+    let x: [i32; N];
+    // Used as an expression.
+    println!("{}", N * 2);
+}
+
+// Used as a field of a struct.
+struct Foo<const N: usize>([i32; N]);
+
+impl<const N: usize> Foo<N> {
+    // Used as an associated constant.
+    const CONST: usize = N * 4;
+}
+
+trait Trait {
+    type Output;
+}
+
+impl<const N: usize> Trait for Foo<N> {
+    // Used as an associated type.
+    type Output = [i32; N];
+}
+```
+
+```rust,compile_fail
+// Examples where const generic parameters cannot be used.
+fn foo<const N: usize>() {
+    // Cannot use in item definitions within a function body.
+    const BAD_CONST: [usize; N] = [1; N];
+    static BAD_STATIC: [usize; N] = [1; N];
+    fn inner(bad_arg: [usize; N]) {
+        let bad_value = N * 2;
+    }
+    type BadAlias = [usize; N];
+    struct BadStruct([usize; N]);
+}
+```
+
+As a further restriction, const parameters may only appear as a standalone
+argument inside of [types] and [array repeat expressions]. In those contexts,
+they may only be used as a single segment [path expression], possibly inside a
+[block] (such as `N` or `{N}`). That is, they cannot be combined with other
+expressions.
 
 ```rust,compile_fail
-// ok: standalone argument
-fn foo<const N: usize>() -> [u8; N] { todo!() }
+// Examples where const parameters may not be used.
+
+// Not allowed to combine in other expressions in types, such as the
+// arithmetic expression in the return type here.
+fn bad_function<const N: usize>() -> [u8; {N + 1}] {
+    // Similarly not allowed for array repeat expressions.
+    [1; {N + 1}]
+}
+```
+
+A const argument in a [path] specifies the const value to use for that item.
+The argument must be a [const expression] of the type ascribed to the const
+parameter. The const expression must be a [block expression][block]
+(surrounded with braces) unless it is a single path segment (an [IDENTIFIER])
+or a [literal] (with a possibly leading `-` token). This syntactic restriction
+is necessary to avoid requiring infinite lookahead when parsing an expression
+inside of a type.
+
+```rust
+fn double<const N: i32>() {
+    println!("doubled: {}", N * 2);
+}
 
-// ERROR: generic const operation
-fn bar<const N: usize>() -> [u8; N + 1] { todo!() }
+const SOME_CONST: i32 = 12;
+
+fn example() {
+    // Example usage of a const argument.
+    double::<9>();
+    double::<-123>();
+    double::<{7 + 8}>();
+    double::<SOME_CONST>();
+    double::<{ SOME_CONST + 5 }>();
+}
+```
+
+When there is ambiguity if a generic argument could be resolved as either a
```

Comment:
This case should talk about different namespaces for clarity, even if we haven't properly documented namespaces (and we can leave a note saying that we haven't yet done so as well).

The way to describe this without mentioning ambiguity would be to say that generic arguments check the type namespace before the value namespace, yes?

Reply:
Hm, that might be tricky, since it would need to also specify which namespace each parameter goes in.  I'd kinda like to tackle that separately if it would be OK?

About a year ago I started on writing an outline of chapters for name resolution, but kinda got overwhelmed and stopped.  I would still like to move it forward.  I was thinking with a rough outline, a good start would just be introducing namespaces. It's hard to document individual pieces in isolation since it is intermingled with scopes and other topics, but I think it can be done incrementally with a plan.  I mainly want to avoid incrementally adding pieces without having a picture of how it should look in the long term, since I think it would otherwise require reorganizing things at each step which can be disruptive and more work.

Reply:
Sure. Opened #924 to discuss this.